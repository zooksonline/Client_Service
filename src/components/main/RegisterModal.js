import React, { useState, useEffect, Fragment } from "react";
import { connect, useSelector } from "react-redux";
import PropTypes from "prop-types";

import { register } from "../../actions/main/registerAction";
import { setDepartment, closeDepartment } from "../../actions/main/depAction";
import { clearErrors } from "../../actions/main/errorAction";

import {
  Alert,
  Button,
  Col,
  Modal,
  ModalHeader,
  ModalBody,
  Row,
  Form,
  FormGroup,
  Label,
  Input,
} from "reactstrap";

const RegisterModal = (props) => {
  const [modal, setmodal] = useState(false);
  const [title, settitle] = useState("นาย");
  const [buasri_id, setbuasri_id] = useState("");
  const [password, setpassword] = useState("");
  const [firstname, setfirstname] = useState("");
  const [lastname, setlastname] = useState("");
  const [email, setemail] = useState("");
  const [dep, setdep] = useState(null);
  const [departments, setdepartments] = useState("");
  const [type, settype] = useState("STAFF");
  const [position] = useState("USER");
  const [active] = useState("ACTIVE");
  const [errorMsg, setErrorMsg] = useState("");

  const isAuthenticated = useSelector(
    (state) => state.main.auth.isAuthenticated
  );
  const listDepartment = useSelector((state) => state.main.departments.list);
  const error = useSelector((state) => state.main.error);

  const ColoredLine = ({ color }) => (
    <hr
      style={{
        color: color,
        backgroundColor: color,
        height: 1,
      }}
    />
  );
  RegisterModal.prototypes = {
    setDepartment: PropTypes.func.isRequired,
    closeDepartment: PropTypes.func.isRequired,
    register: PropTypes.func.isRequired,
    clearErrors: PropTypes.func.isRequired,
  };
  const { setDepartment, closeDepartment, register, clearErrors } = props;

  useEffect(() => {
    if (listDepartment) {
      const GetDepList = async () => {
        await setdepartments(listDepartment);
      };
      GetDepList();
    }
  }, [listDepartment]);

  useEffect(() => {
    // Check for register error
    if (error.id === "REGISTER_FAIL") {
      setErrorMsg(error.msg.msg);
    } else {
      setErrorMsg(null);
    }
    // eslint-disable-next-line
  }, [error]);

  useEffect(() => {
    // If authenticated, close modal
    if (modal) {
      if (isAuthenticated) {
        const ActionUser = async () => {
          await toggle();
        };
        ActionUser();
      }
    }
    // eslint-disable-next-line
  }, [isAuthenticated]);

  const toggle = () => {
    setmodal(!modal);
    if (modal) {
      closeDepartment();
      clearErrors();
    } else {
      setDepartment();
    }
  };

  const onChange = (e) => {
    const { name, value } = e.target;
    if (name === "buasri_id") {
      setbuasri_id(value);
    } else if (name === "password") {
      setpassword(value);
    } else if (name === "title") {
      settitle(value);
    } else if (name === "firstname") {
      setfirstname(value);
    } else if (name === "lastname") {
      setlastname(value);
    } else if (name === "email") {
      setemail(value);
    } else if (name === "type") {
      settype(value);
    } else if (name === "dep") {
      setdep(value);
    }
  };

  const onSubmit = (e) => {
    e.preventDefault();

    const newUser = {
      buasri_id: buasri_id,
      password: password,
      title: title,
      firstname: firstname,
      lastname: lastname,
      email: email,
      dep: dep,
      position: position,
      type: type,
      active: active,
    };

    register(newUser);
  };
  return (
    <Fragment>
      <span onClick={toggle}>
        ลงทะเบียนเข้าใช้งาน <br />
        (สำหรับบุคลากรและนิสิต)
      </span>
      <Modal isOpen={modal} toggle={toggle}>
        <ModalHeader toggle={toggle}>โปรดใส่รายละเอียดผู้ใช้งาน</ModalHeader>
        <ModalBody>
          {errorMsg ? <Alert color="danger">{errorMsg}</Alert> : null}
          <Form onSubmit={onSubmit}>
            <FormGroup>
              <Label for="buasri_id">Buasri ID</Label>
              <Input
                type="text"
                name="buasri_id"
                placeholder="บัวศรีไอดี"
                onChange={onChange}
                required
              ></Input>
            </FormGroup>
            <FormGroup>
              <Label for="password">Password</Label>
              <Input
                type="password"
                name="password"
                placeholder="รหัสผ่าน"
                onChange={onChange}
                required
              ></Input>
            </FormGroup>
            <ColoredLine color="grey" />
            <Row form>
              <Col md={12}>
                <Label for="name">ชื่อ - นามสกุล</Label>
              </Col>
              <Col md={3}>
                <FormGroup>
                  <Input type="select" name="title" onChange={onChange}>
                    <option>นาย</option>
                    <option>นางสาว</option>
                    <option>นาง</option>
                    <option>ดร.</option>
                    <option>ผศ.ดร.</option>
                    <option>รศ.ดร.</option>
                  </Input>
                </FormGroup>
              </Col>
              <Col md={4}>
                <FormGroup>
                  <Input
                    type="text"
                    name="firstname"
                    placeholder="ชื่อ"
                    onChange={onChange}
                    required
                  ></Input>
                </FormGroup>
              </Col>
              <Col md={5}>
                <FormGroup>
                  <Input
                    type="text"
                    name="lastname"
                    placeholder="นามสกุล"
                    onChange={onChange}
                    required
                  ></Input>
                </FormGroup>
              </Col>
            </Row>
            <FormGroup>
              <Label for="email">E-mail</Label>
              <Input
                type="text"
                name="email"
                placeholder="อีเมลที่ใช้ติดต่อ"
                onChange={onChange}
                required
              ></Input>
            </FormGroup>
            <FormGroup>
              <Label for="dep">หน่วยงานที่สังกัด</Label>
              <Input type="select" name="dep" onChange={onChange} required>
                <option value="">โปรดเลือก</option>
                {departments &&
                  departments.map(({ _id, currentNameTH, currentNameEN }) => (
                    <option key={_id} value={currentNameEN}>
                      {currentNameTH}
                    </option>
                  ))}
              </Input>
            </FormGroup>
            <FormGroup>
              <Label for="type">ประเภทผู้ใช้งาน</Label>
              <Input type="select" name="type" onChange={onChange}>
                <option value="STAFF">เจ้าหน้าที่</option>
                <option value="STUDENT">นิสิต</option>
                <option value="PROFESSOR">อาจารย์</option>
              </Input>
            </FormGroup>
            <Button color="dark" style={{ marginTop: "2rem" }} block>
              เพิ่มผู้ใช้งาน
            </Button>
          </Form>
        </ModalBody>
      </Modal>
    </Fragment>
  );
};

export default connect(null, {
  setDepartment,
  closeDepartment,
  register,
  clearErrors,
})(RegisterModal);
