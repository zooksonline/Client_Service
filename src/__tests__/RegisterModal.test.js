import React from "react";
import RegisterModal from "../components/main/RegisterModalV2";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { shallow, mount } from "enzyme";
import "../setupTests";
import renderer from "react-test-renderer";
import shallowRenderer from "react-test-renderer/shallow";

// TypeError: Cannot read property 'getState' of undefined
import { Provider } from "react-redux";
import { createStore, applyMiddleware } from "redux";
const store1 = createStore(() => [], {}, applyMiddleware());

it("Snapshot", () => {
  const renderer = new shallowRenderer();
  const snapshot = renderer.render(<RegisterModal />);
  expect(snapshot).toMatchSnapshot();
});
it("Check Tag", () => {
  // const calculator = shallow(<RegisterModal store={store1} />);
  //check ใน tag form
  // const form = calculator.find("Form");
  // calculator.children().dive().find('FormGroup')
  // expect(calculator.length).toBe(1);
  const wrapper = shallow(<RegisterModal store={store1} />);
  // expect(wrapper.find('.in-bar')).to.have.lengthOf(0);
  expect(wrapper.find("Form"));
});

// it('Test click event', () => {
//   const mockCallBack = sinon.spy();
//   const button = shallow((<RegisterModal onClick={mockCallBack} />));
//   button.find('.onSubmit').simulate('click');
//     expect(mockCallBack).toHaveProperty('callCount', 1);
// });

// it("should call onChange prop with input value", () => {
//   const testState = { name: "buasri_id", faculty: "science" };
//   const wrapper = shallow(
//     <RegisterModal
//       name={testState.name}
//       faculty={testState.faculty}
//       onChange={(e) => {
//         testState[e.target.name] = e.target.value;
//       }}
//     />
//   );
//   wrapper
//     .find(".eiei")
//     .at(0)
//     .simulate("change", { target: { name: "name", value: 50 } });
//   expect(testState.value).toEqual(50);
// });
// const component = shallow(<RegisterModal  />);
// component.find('.eiei').simulate('change');
// expect(component).toBe("Nuttiwut")
// it("Simulate onClick textbox", () => {
//   const calculator = shallow(<RegisterModal />)
//   // กำหนดค่า 1 และ 2
//   const form = calculator.find('Form')
//   // กดปุ่มบวก
//   form.find('.formSubmit').simulate('click')
//   // ผลการบวกต้องเป็น 3
//   expect(form).toHaveBeenCalled()
// });
//   /* ทำการ mock handleTextbox ตัวจริงซะ เพราะเราไม่ได้อยากให้มันรันจริงๆ
//    * เราแค่อยากรู้ว่ามันรัน 1 ครั้งจริงๆมั้ย ถ้าเรา change Input
//    * ใน jest.fn() จะสามารถถามมันได้ว่า คุณรัน 1 ครั้งมั้ย  โดยที่ไม่ต้องไปรันจริง (jest.fn() จะมาทำหน้่าที่แทนฟังก์ชั่นจริง)
//    */
//   // const onSearchMock = jest.fn();
//   // const wrapper = shallow(<RegisterModal/>)
//   // wrapper.find('Form').simulate('click')

//   // //ถาม jest.fn() ที่เรา set ให้มันเมื่อกี๊ว่า ถูกรัน 1 ครั้งจริงๆมั้ย
//   // //การจะใช้ toHaveBeenCalledTimes() ได้ ตัวใน expect ต้องเป็น jest.fn() เท่านั้นนะครับ อันนี้เป็นข้อควรระวัง
//   // expect(wrapper).toHaveBeenCalledTimes(1)
// });