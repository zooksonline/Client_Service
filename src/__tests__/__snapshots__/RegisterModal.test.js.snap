// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
<React.Fragment>
  <span
    onClick={[Function]}
  >
    ลงทะเบียนเข้าใช้งาน 
    <br />
    (สำหรับบุคลากรและนิสิต)
  </span>
  <Modal
    autoFocus={true}
    backdrop={true}
    backdropTransition={
      Object {
        "mountOnEnter": true,
        "timeout": 150,
      }
    }
    centered={false}
    container="body"
    fade={true}
    isOpen={false}
    keyboard={true}
    modalTransition={
      Object {
        "timeout": 300,
      }
    }
    onClosed={[Function]}
    onOpened={[Function]}
    returnFocusAfterClose={true}
    role="dialog"
    scrollable={false}
    toggle={[Function]}
    trapFocus={false}
    unmountOnClose={true}
    zIndex={1050}
  >
    <ModalHeader
      charCode={215}
      closeAriaLabel="Close"
      tag="h5"
      toggle={[Function]}
      wrapTag="div"
    >
      โปรดใส่รายละเอียดผู้ใช้งาน
    </ModalHeader>
    <ModalBody
      tag="div"
    >
      <Form
        className="onSubmit"
        onSubmit={[Function]}
        tag="form"
      >
        <FormGroup
          tag="div"
        >
          <Label
            for="buasri_id"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            Buasri ID
          </Label>
          <Input
            className="eiei"
            name="buasri_id"
            onChange={[Function]}
            placeholder="บัวศรีไอดี"
            required={true}
            type="text"
          />
        </FormGroup>
        <FormGroup
          tag="div"
        >
          <Label
            for="password"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            Password
          </Label>
          <Input
            name="password"
            onChange={[Function]}
            placeholder="รหัสผ่าน"
            required={true}
            type="password"
          />
        </FormGroup>
        <Row
          form={true}
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <Col
            md={12}
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            <Label
              for="name"
              tag="label"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              ชื่อ - นามสกุล
            </Label>
          </Col>
          <Col
            md={3}
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            <FormGroup
              tag="div"
            >
              <Input
                name="title"
                onChange={[Function]}
                type="select"
              >
                <option>
                  นาย
                </option>
                <option>
                  นางสาว
                </option>
                <option>
                  นาง
                </option>
                <option>
                  ดร.
                </option>
                <option>
                  ผศ.ดร.
                </option>
                <option>
                  รศ.ดร.
                </option>
              </Input>
            </FormGroup>
          </Col>
          <Col
            md={4}
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            <FormGroup
              tag="div"
            >
              <Input
                name="firstname"
                onChange={[Function]}
                placeholder="ชื่อ"
                required={true}
                type="text"
              />
            </FormGroup>
          </Col>
          <Col
            md={5}
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            <FormGroup
              tag="div"
            >
              <Input
                name="lastname"
                onChange={[Function]}
                placeholder="นามสกุล"
                required={true}
                type="text"
              />
            </FormGroup>
          </Col>
        </Row>
        <FormGroup
          tag="div"
        >
          <Label
            for="email"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            E-mail
          </Label>
          <Input
            name="email"
            onChange={[Function]}
            placeholder="อีเมลที่ใช้ติดต่อ"
            required={true}
            type="text"
          />
        </FormGroup>
        <FormGroup
          tag="div"
        >
          <Label
            for="dep"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            หน่วยงานที่สังกัด
          </Label>
          <Input
            name="dep"
            onChange={[Function]}
            required={true}
            type="select"
          >
            <option
              value=""
            >
              โปรดเลือก
            </option>
            
          </Input>
        </FormGroup>
        <FormGroup
          tag="div"
        >
          <Label
            for="type"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            ประเภทผู้ใช้งาน
          </Label>
          <Input
            name="type"
            onChange={[Function]}
            type="select"
          >
            <option
              value="STAFF"
            >
              เจ้าหน้าที่
            </option>
            <option
              value="STUDENT"
            >
              นิสิต
            </option>
            <option
              value="PROFESSOR"
            >
              อาจารย์
            </option>
          </Input>
        </FormGroup>
      </Form>
    </ModalBody>
  </Modal>
</React.Fragment>
`;
